Sofa/Component/Constraint/Lagrangian/Solver/src/sofa/component/constraint/lagrangian/solver/ConstraintForceExporter.h

```diff
@@ -0,0 +1,111 @@
+* This program is free software; you can redistribute it and/or modify it     *
+* under the terms of the GNU Lesser General Public License as published by    *
+* the Free Software Foundation; either version 2.1 of the License, or (at     *
+* your option) any later version.                                             *
+*                                                                             *
+* This program is distributed in the hope that it will be useful, but WITHOUT *
+* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or       *
+* FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License *
+* for more details.                                                           *
+*                                                                             *
+* You should have received a copy of the GNU Lesser General Public License    *
+* along with this program. If not, see <http://www.gnu.org/licenses/>.        *
+*******************************************************************************
+* Authors: The SOFA Team and external contributors (see Authors.txt)          *
+*                                                                             *
+* Contact information: contact@sofa-framework.org                             *
+******************************************************************************/
+#pragma once
+#include <sofa/component/constraint/lagrangian/solver/config.h>
+
+#include <sofa/component/constraint/lagrangian/solver/ConstraintSolverImpl.h>
+#include <sofa/core/behavior/MechanicalState.h>
+
+#include <sofa/core/fwd.h>
+
+namespace sofa::component::constraint::lagrangian::solver
+{
+
+/**
+* Small utility component to compute forces (and wrenches in case of rigids)
+* from constraints computed by a constraint solver on a mechanical state,
+* at the end of each animation step.
+* Can also draw the forces for debug purpose (but it does not draw the wrenches if applicable).
+*/
+template <typename DataTypes>
+class ConstraintForceExporter : public sofa::core::objectmodel::BaseObject
+{
+public:
+    SOFA_CLASS(ConstraintForceExporter, sofa::core::objectmodel::BaseObject);
+
+    using Inherited = sofa::core::objectmodel::BaseObject;
+    using Deriv = typename DataTypes::Deriv;
+    using VecDeriv = typename DataTypes::VecDeriv;
+
+protected:
+    ConstraintForceExporter();
+    ~ConstraintForceExporter() override = default;
+public:
+    void init() override;
+    void handleEvent(sofa::core::objectmodel::Event* event) override;
+    void draw(const sofa::core::visual::VisualParams* vparams) override;
+
+    Data<VecDeriv> d_constraintForces; ///< (output) Forces (or wrenches if rigids) computed from the mechanical state using the constraint solver.
+    Data<bool> d_draw; ///< (debug) draw forces (as an arrow) for each position of the mechanical object.
+    Data<double> d_drawForceScale; ///< (debug) Scale to apply on the force (draw as an arrow).
```

Comment:
? d_draw and d_drawForceScale are already public
You meant let them public and set d_constraintForces private ?

Reply:
you're right. But yes, since `d_constraintForces` is an output, it should not be modified by anyone. So I suggest to make it protected.

Reply:
In the past it was considered that as every data is "public" (as it can be founded with findData in any case) and thus it was ok to put everything in public. Do you think we should change that ?

Reply:
I would say it was for consistency. As they are public using the findData mecanism and that this "publicity" propagates up to the python API and in the GUI... there was no real reason for hiding that in c++.

In rare case the Data are not visible,  when they are flagged to be hidden in the GUI or if they are not registered into the Base::datas (so not using initData()) then they can be considered as  "private". 

For these ones... it may be questionable about why the variable is in a data and not juste a raw member;